File: July13/CIRKILL1.c
#include<stdio.h>
#include<string.h>
#include<math.h>
int arr1[30];
int arr2[30];
int check(int pos1,int pos2,int pos3,int pos4){
	if(arr2[pos1]==arr2[pos2] && arr2[pos2]==arr2[pos3]){
		return 0;
	}
	if(arr1[pos1]==arr1[pos2] && arr1[pos1]==arr1[pos3]){
		return 0;
	}
	double m1,m2;
	double x,y;
	double p,q,r,s;
	p=(arr1[pos1]+arr1[pos2])*1.0/2;
	q=(arr2[pos1]+arr2[pos2])*1.0/2;
	r=(arr1[pos2]+arr1[pos3])*1.0/2;
	s=(arr2[pos2]+arr2[pos3])*1.0/2;
	if(arr1[pos1]==arr1[pos2]){
		y=(arr2[pos1]+arr2[pos2])*1.0/2;
		if(arr2[pos2]==arr2[pos3]){
			x=(arr1[pos2]+arr1[pos3])*1.0/2;
		}
		else{
			m2=(arr2[pos2]-arr2[pos3])*1.0/(arr1[pos2]-arr1[pos3]);
			x=r-m2*1.0*(y-s);
		}
	}	
	else if(arr1[pos2]==arr1[pos3]){
		y=(arr2[pos2]+arr2[pos3])*1.0/2;	
		if(arr2[pos1]==arr2[pos2]){
			x=(arr1[pos1]+arr1[pos2])*1.0/2;
		}
		else{
			m1=(arr2[pos2]-arr2[pos1])*1.0/(arr1[pos2]-arr1[pos1]);
			x=p-m1*1.0*(y-q);
		}
	}
	else{
		m1=(arr2[pos2]-arr2[pos1])*1.0/(arr1[pos2]-arr1[pos1]);
		m2=(arr2[pos2]-arr2[pos3])*1.0/(arr1[pos2]-arr1[pos3]);
		x=((r*1.0/m2)-(p*1.0/m1)+s-q)*m1*m2/(m1-m2);
		y=-1.0/m1*x+p*1.0/m1+q;
	}
	float rad=sqrt(pow((x-arr1[pos1]),2)+pow((y-arr2[pos1]),2));
	float distance=sqrt(pow((x-arr1[pos4]),2)+pow((y-arr2[pos4]),2));
	if(distance<=rad){
		//printf("Returning 1 for %d %d %d %d for rad = %f and distance = %f and x=%f and y=%f\n",pos1,pos2,pos3,pos4,rad,distance,x,y);
		return 1;
	}
	return 0;
}
int main(){
	int T;
	int N;
	int i,j,k,l;
	int total;
	int count;
	scanf("%d",&T);
	while(T>0){
		count=0;
		total=0;
		scanf("%d",&N);
		total=N*(N-1)*(N-2)*(N-3)/6;
		memset(arr1,0,sizeof(arr1));
		memset(arr2,0,sizeof(arr2));
		i=0;
		while(i<N){
			scanf("%d %d",&arr1[i],&arr2[i]);
			i++;
		}
		i=0;
		while(i<(N-2)){
			j=i+1;
			while(j<(N-1)){
				k=j+1;
				while(k<(N)){
					l=0;
					while(l<N){
						if(!(l==i || l==j || l==k)){
							if(check(i,j,k,l)){
								count++;
							}
						}
						l++;
					}
					k++;
				}
				j++;
			}
			i++;
		}
		//printf("Count is %d\n",count);
		printf("%.6f\n",((1.0*count)/total));
		T--;
	}
	return 0;
}

File: July13/CIRKILL1.java
import java.io.*;
class Helper{
	private int arr1[];
	private int arr2[];
	private int size;
	private int i,j,k,l;
	String str[];
	public Helper(int n,BufferedReader I)throws IOException{
		size=n;
		arr1=new int[size];
		arr2=new int[size];
		i=0;
		while(i<size){
			str=(I.readLine()).split(" ");
			arr1[i]=Integer.parseInt(str[0]);
			arr2[i]=Integer.parseInt(str[1]);
			i++;
		}
	}
	private boolean check(int i1,int i2,int i3,int i4){
		double A1,A2,B1,B2,C1,C2;
		double p,q,r,s;
		double x,y;
		double det;
		double rad,distance;
		if((arr1[i1]==arr1[i2] && arr1[i2]==arr1[i3])|| (arr2[i1]==arr2[i2] && arr2[i2]==arr2[i3])){
			return false;
		}
		else{
			A1=arr2[i2]-arr2[i1];
			B1=arr1[i1]-arr1[i2];
			A2=arr2[i2]-arr2[i3];
			B2=arr1[i3]-arr1[i2];
			p=(arr1[i1]+arr1[i2])*1.0;
			q=(arr2[i1]+arr2[i2])*1.0;
			r=(arr1[i2]+arr1[i3])*1.0;
			s=(arr2[i2]+arr2[i3])*1.0;
			C1=A1*q-B1*p;
			C2=A2*s-B2*r;
			det=A1*B2-A2*B1;
			if(det==0){
				return false;
			}
			else{
				x=(A2*C1-A1*C2)*1.0;
				y=(B2*C1-B1*C2)*1.0;
			}
		}
		rad=Math.sqrt(Math.pow((x-2.0*det*arr1[i2]),2)+Math.pow((y-2.0*det*arr2[i2]),2));
		distance=Math.sqrt(Math.pow((x-2.0*det*arr1[i4]),2)+Math.pow((y-2.0*det*arr2[i4]),2));
		if(distance<=rad){
			//System.out.println("For "+rad+" "+distance+" "+i1+" "+i2+" "+i3+" "+i4);
			return true;
		}
		return false;
	}
	public int count(){
		int Count=0;
		i=0;
		while(i<(size-2)){
			j=i+1;
			while(j<(size-1)){
				k=j+1;
				while(k<size){
					l=0;
					while(l<size){
						if(!(l==i || l==j || l==k)){
							if(check(i,j,k,l)){
								Count++;
							}
						}
						l++;
					}
					k++;
				}
				j++;
			}
			i++;
		}
		return Count;
	}
}
class CIRKILL1{
	public static void main(String args[])throws IOException{
		Helper ob;
		BufferedReader I=new BufferedReader(new InputStreamReader(System.in));
		int T=Integer.parseInt(I.readLine());
		int N;
		double Total;
		while(T>0){
			N=Integer.parseInt(I.readLine());
			ob=new Helper(N,I);
			Total=N*(N-1)*(N-2)*(N-3)*1.0/6;
			System.out.println(ob.count()*1.0/Total);
			T--;
		}
	}
}
